llvm/lib/Transforms/Utils/BuildLibCalls.cpp

```diff
@@ -279,6 +279,9 @@ bool llvm::inferNonMandatoryLibFuncAttrs(Function &F,
     Changed |= setNonLazyBind(F);
 
   switch (TheLibFunc) {
+  case LibFunc_nan:
+  case LibFunc_nanf:
+  case LibFunc_nanl:
```

Comment:
Is this correct? These are spec'd to call strtod and friends, which set errno.

Reply:
> Error handling
This function is not subject to any of the error conditions specified in [math_errhandling](https://en.cppreference.com/w/c/numeric/math/math_errhandling).

Reply:
LLVM's libc implementation sets errno: https://github.com/llvm/llvm-project/pull/76690/files

Reply:
glibc also sets errno.

Reply:
> Is this correct? These are spec'd to call strtod and friends, which set errno.

In this case, strtoX never sets ERANGE. So it looks ok.

Reply:
```
#include <math.h>
#include <errno.h>
#include <stdio.h>

int main() {
	errno = 0;
	double d = nan("99999999999999999999999999999999999999999");
	int err = errno;
	printf("%d %f\n", err, d);
	return 0;
}

$ clang -lm test.c && ./a.out 
34 nan
```
With glibc. Is this a glibc bug?

Reply:
The conditions where strtod would hit an error shouldn't apply for nan though?

Reply:
`strotod` says for NaN sequences:
>  A character sequence `NAN` or `NAN(n-char-sequenceopt)` is interpreted as a quiet NaN, if supported in the return type, else like a subject sequence part that does not have the expected form; the meaning of the n-char sequence
is implementation-defined.

It gives explicit rules for overflow and underflow, which can set `errno` values, but is silent on what happens with the n-char sequence being out of range of the payload, outside of the above quote. But... there's not really any latitude I see in the specification for `errno` to be set (unless you think 7.12.1 applies to `nan` and sNaN is a domain error).

glibc fails because it's using `strtoull` here, and `strtoull` does `ERANGE` on overflow. But I think this is a glibc bug, it shouldn't be setting errno.

(I'm not fully conversant in llvm-libc implementation, but I think the llvm-libc implementation doesn't ever set `errno` in `nan`--the interface allows it to, but nothing appears to set an error).

Reply:
I filed a glibc bug here: https://sourceware.org/bugzilla/show_bug.cgi?id=32045